### PR TITLE
Rename classes, tags and functions

### PR DESCRIPTION
**Pull request by murtukov**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Documented?   | yes
| Fixed tickets | #775 

Implementation of #775

List of renames:

- Class `GlobalVariables` -> `GraphQLServices`
- Class `ResolverResolver` -> `QueryResolver`
- Interface `ResolverInterface` -> `QueryInterface`
- Service tag `overblog_graphql.global_variable` -> `overblog_graphql.service`
- Service tag `overblog_graphql.resolver` -> `overblog_graphql.query`
- Expression function `resolver` -> `query`

Also all UPGRADE-files for previous versions have been removed.

### CODE REVIEW

File: UPGRADE-1.0.md [hunk 1]
```diff
@@ -4,8 +4,10 @@ UPGRADE FROM 0.13 to 1.0
 # Table of Contents
 
 - [Customize the cursor encoder of the edges of a connection](#customize-the-cursor-encoder-of-the-edges-of-a-connection)
-- [Change arguments of `TypeGenerator`](#change-arguments-of-typegenerator)
+- [Change arguments of `TypeGenerator`](#change-arguments-of-typegenerator-class)
 - [Add magic `__get` method to `ArgumentInterface` implementors](#add-magic-__get-method-to-argumentinterface-implementors)
+- [Rename `GlobalVariables` to `GraphQLServices`](#)
```
Comment (mcg-web) [inline UPGRADE-1.0.md:9]: missing links

File: UPGRADE-1.0.md [hunk 2]
```diff
@@ -4,8 +4,10 @@ UPGRADE FROM 0.13 to 1.0
...
+- [Change `overblog_graphql.global_variable` tag]()
```
Comment (mcg-web) [inline UPGRADE-1.0.md:11]: Missing parts for resolver, mutation, all public tags, interfaces, classes ... changes (just a reminder)

File: docs/definitions/expression-language.md [hunk 3]
```diff
@@ -320,11 +320,11 @@ Examples
 @=getUser().firstName === 'adam'
 ```
 
+
 ## Registered variables:
 
 | Variable             | Description  | Scope|
 |:-------------------- |:------------ |:---- |
-| `typeResolver`       | An object of class `Overblog\GraphQLBundle\Resolver\TypeResolver`| global|
```
Comment (mcg-web) [inline docs/definitions/expression-language.md:327]: why this change ?

**Review by mcg-web — CHANGES_REQUESTED** [review]
All these changes make sense but also make the migration from `0.13` to `1.0` a very heavy task. We should maybe introduce all these in a `0.14` version with deprecations so it ease the process.
What do you think of this ?

**Comment by murtukov** [comment]
@mcg-web I think making 0.14 as a transition version is a good idea.

File: docs/definitions/expression-language.md (hunk 3 above)
Comment (murtukov) [reply docs/definitions/expression-language.md:327]: @mcg-web It's a mistake. First I removed `typeResolver` from expressions, then I rerolled this change, but forgot to reroll changes done to the documentation. I'll fix this.

File: UPGRADE-1.0.md (hunk 2 above)
Comment (murtukov) [reply UPGRADE-1.0.md:11]: @mcg-web same here, I'll write the UPDATE documentation.

File: UPGRADE-1.0.md (hunk 1 above)
Comment (murtukov) [reply UPGRADE-1.0.md:9]: @mcg-web Yep, I will write the UPDATE documentation.

**Comment by murtukov** [comment]
@mcg-web I made it BC. Also I removed all UPGRADE files for previous versions. We can have a single UPGRADE.md file in each branch, like `webonyx/graphql-php` does.

**Review by Vincz — APPROVED** [review]

File: src/Definition/Resolver/ResolverInterface.php [hunk 4]
```diff
@@ -4,6 +4,10 @@
 
 namespace Overblog\GraphQLBundle\Definition\Resolver;
 
+/**
```
Comment (mcg-web) [inline src/Definition/Resolver/ResolverInterface.php:7]: also add a `trigger_error` call please

File: src/DependencyInjection/Compiler/ResolverTaggedServiceMappingPass.php [hunk 5]
```diff
@@ -8,6 +8,12 @@
 use function is_string;
 use function sprintf;
 
+/**
```
Comment (mcg-web) [inline src/DependencyInjection/Compiler/ResolverTaggedServiceMappingPass.php:11]: also add a `trigger_error` call please

File: src/ExpressionLanguage/ExpressionFunction/GraphQL/Mutation.php [hunk 6]
```diff
@@ -8,11 +8,32 @@
...
+    public function __construct($name = self::NAME)
     {
         parent::__construct(
             $name,
-            fn ($alias, $args = '[]') => "$this->globalVars->get('mutationResolver')->resolve([$alias, $args])"
+            function (string $alias, ...$args) {
+                $count = count($args);
+
+                // TODO: remove the following if-else-block in 1.0
+                if (1 === $count && '$' !== $args[0][0]) {
```
Comment (mcg-web) [inline src/ExpressionLanguage/ExpressionFunction/GraphQL/Mutation.php:21]: `$args[0][0]` can be unset, so should we not test it to avoid notice ?

**Review by mcg-web — CHANGES_REQUESTED** [review]
We should also keep some legacy tests.

File: src/Definition/Resolver/ResolverInterface.php (hunk 4 above)
Comment (murtukov) [reply src/Definition/Resolver/ResolverInterface.php:7]: How do I add `trigger_error` in an interface?

Comment (mcg-web) [reply src/Definition/Resolver/ResolverInterface.php:7]: You should just added before the class or the interface definition.
You can see examples here:
https://symfony.com/doc/current/contributing/code/conventions.html

Comment (murtukov) [reply src/Definition/Resolver/ResolverInterface.php:13]: Done

File: src/DependencyInjection/Compiler/ResolverTaggedServiceMappingPass.php (hunk 5 above)
Comment (murtukov) [reply src/DependencyInjection/Compiler/ResolverTaggedServiceMappingPass.php:15]: done

File: src/ExpressionLanguage/ExpressionFunction/GraphQL/Mutation.php (hunk 6 above)
Comment (murtukov) [reply src/ExpressionLanguage/ExpressionFunction/GraphQL/Mutation.php:21]: `$args[0][0]` can never be unset. The `1 === $count` makes sure that `$args[0]` is set. Furthermore all items in the `$args` array are always non-empty strings, which means `$args[0][0]` points to the first letter of the string and is never unset.

Now a little explanation, why I did this check. We are changing the signature of the `mutation` function.

old: `mutation(string $alias, array $args = [])`
new: `mutation(string $alias, ...$args)`

The` if-block` tries to guess what signature user currently tries to use. If he calls the function with only 1 argument, which is an array (`'[' === $args[0][0]`), then it considers it old signature call, otherwise it uses the new signature.

The only drawback here is the case, where user tries to use the new signature and sends 1 array-argument:
`mutation("my_resolver", ['name', 'lastname'])`
So the bundle will think it's and old signature and will try to send `name` and `lastname` as 2 separate arguments to the resolver.

**Comment by murtukov** [comment]
@mcg-web I tweaked some some tests to use legacy code, so there are no `@codeCoverageIgnore` tags required anymore

File: src/ExpressionLanguage/ExpressionFunction/GraphQL/Mutation.php (hunk 6 above)
Comment (mcg-web) [reply src/ExpressionLanguage/ExpressionFunction/GraphQL/Mutation.php]: yes I agree, I don't see any easy way to make the diffence between old and new signature in this case without using reflection.

Comment (murtukov) [reply src/ExpressionLanguage/ExpressionFunction/GraphQL/Mutation.php]: @mcg-web so what do you think I should do? Should I leave it as is?

**Comment by murtukov** [comment]
@mcg-web 

I added some additional checks with reflection to decide, which signature should be used.

Here are some examples of which signature is used by certain calls:

1. `mutation('get_posts')` - **new signature**
2. `mutation('get_posts', 'hello')` - **new signature**
3. `mutation('get_posts', [args, 'test'])`:
    - If the array contains the same number of elements as the number of arguments in proxy resolver: **old signature**
    - Otherwise: **new signature**

**Review by mcg-web — APPROVED** [review]
